cluster-endpoints/src/grpc_multiplex.rs

```diff
@@ -68,28 +77,75 @@ pub fn create_grpc_multiplex_blocks_subscription(
-        (multiplex_stream_confirmed, multiplex_stream_finalized).merge();
-
-    let (multiplexed_finalized_blocks, jh_channelizer) =
-        channelize_stream(merged_stream_confirmed_finalize);
+    // return value is the broadcast receiver
+    let (producedblock_sender, blocks_output_stream) =
+        tokio::sync::broadcast::channel::<ProducedBlock>(1000);
+
+    let jh_block_emitter_task = {
+        tokio::task::spawn(async move {
+            // by blockhash
+            let mut recent_confirmed_blocks = HashMap::<String, ProducedBlock>::new();
+            let mut confirmed_blocks_stream = std::pin::pin!(confirmed_blocks_stream);
+            let mut finalized_blockmeta_stream = std::pin::pin!(finalized_blockmeta_stream);
+
+            let sender = producedblock_sender;
+            let mut cleanup_tick = tokio::time::interval(Duration::from_secs(5));
+            let mut current_slot: Slot = 0;
+            loop {
+                tokio::select! {
+                    confirmed_block = confirmed_blocks_stream.next() => {
+                        let confirmed_block = confirmed_block.expect("confirmed block from stream");
+                        current_slot = confirmed_block.slot;
+                        trace!("got confirmed block {} with blockhash {}",
+                            confirmed_block.slot, confirmed_block.blockhash.clone());
+                        if let Err(e) = sender.send(confirmed_block.clone()) {
+                            panic!("Confirmed block stream send gave error {e:?}");
+                        }
+                        recent_confirmed_blocks.insert(confirmed_block.blockhash.clone(), confirmed_block);
+                    },
+                    meta_finalized = finalized_blockmeta_stream.next() => {
+                        let blockhash = meta_finalized.expect("finalized block meta from stream");
+                        if let Some(cached_confirmed_block) = recent_confirmed_blocks.remove(&blockhash) {
+                            let finalized_block = cached_confirmed_block.to_finalized_block();
+                            debug!("got finalized blockmeta {} with blockhash {}",
+                                finalized_block.slot, finalized_block.blockhash.clone());
+                            if let Err(e) = sender.send(finalized_block) {
+                                panic!("Finalized block stream send gave error {e:?}");
+                            }
+                        } else {
+                            debug!("finalized block meta received for blockhash {} which was never seen or already emitted", blockhash);
+                        }
+                    },
+                    _ = cleanup_tick.tick() => {
+                        let size_before = recent_confirmed_blocks.len();
+                        recent_confirmed_blocks.retain(|_blockhash, block| {
+                            block.slot > current_slot - 100 // must be greater than finalized slot distance (32)
```

Comment:
Not a good idea, In some cases it takes time to finalize like if there are some complicated bank forks. So current_slot should be last_finalized_slot

Reply:
ah! very good point. my logic was very wrong.

Reply:
should be fixed now